File: py/trees/bst.py
from node import Node, NUM_OF_CHARS_PER_VALUE

class BST:
	def __init__(self):
		self.root: Node = None

	def _find(self, value: int) -> tuple[Node, Node]:
		parent = None
		node = self.root

		while node and node.value != value:
			parent = node
			if value < node.value:
				node = node.left
			else:
				node = node.right

		return (parent, node)

	def insert(self, value: int) -> Node:
		if not self.root:
			self.root = Node(value)
			return

		parent, node = self._find(value)
		if node:  # there is already a node with value
			return node

		new_node = Node(value)
		new_node.parent = parent
		if new_node.value < parent.value:
			parent.left = new_node
		else:
			parent.right = new_node

		return new_node

	def search(self, value: int) -> Node:
		_, node = self._find(value)
		return node

	def delete(self, value: int) -> Node:
		found = self.search(value)
		if not found:
			return None

		if found.left and found.right:  # if there are two children
			nextInOrderNode = found.right
			while nextInOrderNode.left:
				nextInOrderNode = nextInOrderNode.left

			# swap found value with the next in order
			found.value = nextInOrderNode.value
			# mark the next in order node to be deleted
			found = nextInOrderNode

		# found node has at maximum one non nil child
		child = found.left if found.left else found.right
		if child:
			child.parent = found.parent

		parent = found.parent
		if not parent:
			self.root = child
			return self.root

		if parent.left == found:
			parent.left = child
		else:
			parent.right = child
		return parent

	def print(self):
		layers = [[self.root]]
		values = []
		sep = [0, NUM_OF_CHARS_PER_VALUE]

		while not all(n is None for n in layers[-1]):
			values.append([str(n.value) if n else "." * NUM_OF_CHARS_PER_VALUE
						   for n in layers[-1]])
			layers.append([m for n in layers[-1] for m in
						   ([n.left, n.right] if n else [None, None])])
			sep.append(sep[-1]*NUM_OF_CHARS_PER_VALUE + NUM_OF_CHARS_PER_VALUE)

		for i, values in enumerate(values):
			print(" "*sep[-3 - i], sep="", end="")
			print((" "*sep[-2 - i]).join(values))

def test_tree(tree: BST):
	print("Empty tree print")
	tree.print()
	print(f"Empty tree search {tree.search(0)}")
	tree.delete(0)
	print("Empty tree delete")

	for i in [60, 40, 30, 70, 80, 50, 10, 20]:
		print(f"insert({i})")
		tree.insert(i)
		tree.print()

	for i in [30, 60, 50, 90]:
		print(f"search({i}): {tree.search(i)}")

	for i in [30, 50, 70, 60]:
		print(f"delete({i})")
		tree.delete(i)
		tree.print()

File: go/trees/bst.go
package main

import (
	"fmt"
	"slices"
	"strings"
)

type BST struct {
	root *Node
}

func NewBST() *BST {
	return &BST{}
}

func (b *BST) find(value int) (*Node, *Node) {
	var parent *Node
	node := b.root

	for node != nil && node.value != value {
		parent = node
		if value < node.value {
			node = node.left
		} else {
			node = node.right
		}
	}
	return parent, node
}

func (b *BST) Insert(value int) *Node {
	if b.root == nil {
		b.root = &Node{value: value}
		return b.root
	}

	parent, node := b.find(value)
	if node != nil { // there is already a node with value
		return node
	}

	new_node := &Node{value: value, parent: parent}
	if value < parent.value {
		parent.left = new_node
	} else {
		parent.right = new_node
	}

	return new_node
}

func (b *BST) Search(value int) *Node {
	_, node := b.find(value)
	return node
}

func (b *BST) Delete(value int) *Node {
	found := b.Search(value)
	if found == nil {
		return nil
	}

	if found.left != nil && found.right != nil { // if there are two children
		nextInOrderNode := found.right
		for nextInOrderNode.left != nil {
			nextInOrderNode = nextInOrderNode.left
		}
		found.value = nextInOrderNode.value // swap found value with the next in order
		found = nextInOrderNode             // mark the next in order node to be deleted
	}

	child := found.right // found node has at maximum one non nil child
	if found.left != nil {
		child = found.left
	}
	if child != nil {
		child.parent = found.parent
	}

	parent := found.parent
	if parent == nil {
		b.root = child
		return b.root
	}

	if parent.left == found {
		parent.left = child
	} else {
		parent.right = child
	}
	return parent
}

func (b *BST) Print() {
	if b.root == nil {
		return
	}

	layer := []*Node{b.root}
	values := [][]string{}
	sep := []int{0, numOfCharsPerValue}

	for slices.IndexFunc(layer, func(n *Node) bool { return n != nil }) != -1 {
		newLayer := []*Node{}
		newValues := []string{}
		for _, n := range layer {
			newValues = append(newValues, n.String())
			newLayer = append(newLayer, n.l(), n.r())
		}

		layer = newLayer
		values = append(values, newValues)
		sep = append(sep, (sep[len(sep)-1]+1)*numOfCharsPerValue)
	}

	for i, vals := range values {
		fmt.Println(strings.Repeat(" ", sep[len(sep)-3-i]) + strings.Join(vals, strings.Repeat(" ", sep[len(sep)-2-i])))
	}
}
